mgmt/src/models/external/overlay/vpcpeering.rs

```diff
@@ -106,9 +291,43 @@ impl VpcPeeringTable {
         self.0.is_empty()
     }
 
+    fn validate_peering_addition(&self, new_peering: &VpcPeering) -> ConfigResult {
+        // Check that exposes in the new peering do not collide with any of the exposes in the
+        // existing peerings in the table, for the same VPCs
+        for vpc in [&new_peering.left.name, &new_peering.right.name] {
+            let (new_local, _) = new_peering.get_peering_manifests(vpc);
+            for peering in self.peerings_vpc(vpc) {
+                let (local, _) = peering.get_peering_manifests(vpc);
+                for new_expose in &new_local.exposes {
+                    for expose in &local.exposes {
+                        validate_overlapping(
+                            &new_expose.ips,
+                            &new_expose.nots,
+                            &expose.ips,
+                            &expose.nots,
+                        )?;
+                        validate_overlapping(
+                            &new_expose.as_range,
+                            &new_expose.not_as,
+                            &expose.as_range,
+                            &expose.not_as,
+                        )?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
     /// Add a [`VpcPeering`] to a [`VpcPeeringTable`]
     pub fn add(&mut self, peering: VpcPeering) -> ConfigResult {
         peering.validate()?;
+        self.validate_peering_addition(&peering)?;
+
+        // First look for an existing entry, to avoid inserting a duplicate peering
```

Comment:
The duplicate check is already performed in line 331.

Reply:
You're absolutely right. I wonder if then we should change the code after the check, in particular wrt to the return value. It should probably succeed always and the Err be unreachable?

Reply:
It should, at least as long as there's only one thread reading from the table. I hesitated to remove the other check, I could mark it as `unreachable!()` so that we fail loudly if the assumption ever turns out to no longer be valid

Reply:
Either option is fine. You could actually remove the extra check you added:
* if everything is ok (no duplicate), the code will just do an insertion (instead of lookup + instertion)
* in the bad case (duplication), true, the insertion will overwrite the previous element. However, we'll be discarding the config. So, the most optimal option in terms of code and performance may be just inserting and complaining afterwards.

Reply:
Unless you're opposed to it I think I prefer to keep the check.

I understand we currently reject the config on error, but we could imagine someone changing or reusing the code elsewhere and do something like “try to insert this rule, else fallback to that...” and we'll get the rule inserted anyway as a side effect. That's the case in my tests by the way, where I try to insert several rules successively, sometimes expecting an error, but not expecting rules to be inserted in case of error.

Performance-wise, you're correct, although I think the difference is minor, and this happens at config time not in the hot path, so it shouldn't matter.